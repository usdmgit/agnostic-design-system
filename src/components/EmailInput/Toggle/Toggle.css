@import '../../styles/theme/index.css';

.toggle-button span {
  z-index: var(--toggle-span-z-index, 1);
}

.toggle-button[category='neutral']:hover {
  background-color: var(--toggle-background-color-hover);
}

.toggle-button[category='neutral']:active,
.toggle-button[category='neutral']:focus {
  background-color: var(--toggle-background-color-active, var(--structure-color-400));
}

.toggle-button[category='neutral'] {
  background-color: var(--toggle-background-color, var(--structure-color-200));
  transition: var(--toggle-button-transition, all 0.5s ease);
}

.toggle-input {
  position: absolute;
  visibility: hidden;
}

.toggle-wrapper {
  background-color: var(--toggle-wrapper-background-color, var(--structure-color-200));
  border-radius: var(--toggle-wrapper-border-radius, 3px);
  height: var(--toggle-wrapper-height, 50px);
  padding: var(--toggle-padding, 10px);
  white-space: var(--toggle-white-space, normal);
  width: 100%;
}

.toggle {
  background-color: var(--toggle-background-color, var(--structure-color-200));
  border-radius: var(--toggle-border-radius, 3px);
  display: var(--toggle-display, flex);
  height: var(--toggle-height, 50px);
  position: var(--toggle-position, relative);
  width: var(--toggle-width, 100%);
}
/*[data-number-of-items]*/
.toggle-active-btn {
  background-color: var(--toggle-active-btn-background-color, var(--structure-color-white));
  box-shadow: var(--toggle-active-btn-box-shadow, 0 2px 15px rgb(0 0 0 / 15%));
  border-radius: var(--toggle-active-btn-border-radius, 3px);
  cursor: var(--toggle-active-btn-cursor, pointer);
  display: var(--toggle-active-btn-display, block);
  height: var(--toggle-active-btn-height, 50px);
  position: var(--toggle-active-btn-position, absolute);
  transition: var(
    --toggle-active-btn-transition,
    transform 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94)
  );
  z-index: var(--toggle-active-btn-z-index, 1);
}

.toggle-active-btn[data-active-item='false'] {
  transform: var(--toggle-button-transform, translateX(50%));
}

.toggle-label {
  align-items: var(--toggle-label-align-items, center);
  display: var(--toggle-label-display, flex);
  flex-direction: var(--toggle-label-flex-direction, column);
  height: var(--toggle-label-height, 100%);
  margin: var(--toggle-label-margin);
  position: var(--toggle-label-position, relative);
}

.toggle-label[data-active='false'] button {
  color: var(--toggle-color, var(--structure-color-600));
}

.toggle-label[data-active='false'] button:hover {
  color: var(--toggle-color-hover, var(--structure-color-600));
}

.toggle-label[data-active='true'] button,
.toggle-label[data-active='true'] button:hover {
  color: var(--toggle-color-active, var(--primary-color-900));
}
