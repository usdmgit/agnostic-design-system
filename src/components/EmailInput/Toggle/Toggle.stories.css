@import '../../styles/theme/index.css';

.custom-toggle {
  --toggle-wrapper-background-color: blue;
  --toggle-active-btn-background-color: lightblue;
  --toggle-color-active: white;
  --toggle-color-hover: red;
  --toggle-color: purple;
  --toggle-background-color-hover: yellow;
  --toggle-background-color-active: green;
}
