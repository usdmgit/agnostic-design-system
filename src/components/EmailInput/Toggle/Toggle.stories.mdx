# Toggle Component

## Type: Simple

## Toggle Props

|        Prop        |        Type         |                             Description                             | Default | Mandatory |
| :----------------: | :-----------------: | :-----------------------------------------------------------------: | :-----: | :-------: |
|     activeItem     |          T          |              Specifies which item is currently active.              |    -    |   true    |
|       items        |        T[ ]         |    Specifies all the items that should be present on the toggle.    |    -    |   true    |
| variablesClassName |       string        | Receives the class name to add custom styles to the Toggle element. |    -    |   false   |
|      getLabel      | (item: T) => string |               Returns the label for a specific item.                |    -    |   true    |
|      getValue      | (item: T) => string |               Returns the valude for a specific item.               |    -    |   true    |
|      onChange      | (item?: T) => void  |   Specifies what should happen when the toggle option is clicked.   |    -    |   true    |

## Getting Started

```javascript
import React, { useState } from 'react';
import { Toggle } from '@codelittinc/agnostic-design-system';

const Toggle = args => {
  const [active, setActive] = useState(args.activeItem);

  return <Toggle {...args} activeItem={active} onChange={setActive} />;
};

export default Toggle;
```

## CSS Variables Available

The component toggle is accepting style customization through its `variablesClassName` property
`<Toggle variablesClassName={styles['my-toggle-css']} />`

Apart from this, we are also using the component Button, and its style can be changed just as we do on the original component.

### Toggle variables available

|     Property      |    Attribute     |     State     |
| :---------------: | :--------------: | :-----------: |
|       item        | background-color |               |
|      toggle       | background-color | hover, active |
|      toggle       |  border-radius   |               |
|      toggle       |      color       | hover, active |
|      toggle       |     display      |               |
|      toggle       |      height      |               |
|      toggle       |     padding      |               |
|      toggle       |     position     |               |
|      toggle       |    white-space   |               |
|      toggle       |      width       |               |
| toggle-active-btn |  border-radius   |               |
| toggle-active-btn |      height      |               |
| toggle-active-btn |     padding      |               |
| toggle active-btn |     position     |               |
| toggle active-btn | background-color |               |
| toggle active-btn |    box-shadow    |               |
| toggle active-btn |    transition    |               |
| toggle active-btn |      cursor      |               |
| toggle active-btn |     display      |               |
| toggle active-btn |     z-index      |               |
|   toggle-button   |    transform     |               |
|   toggle-button   |    transition    |               |
|   toggle-label    |   align-items    |               |
|   toggle-label    |     display      |               |
|   toggle-label    |  flex-direction  |               |
|   toggle-label    |      height      |               |
|   toggle-label    |      margin      |               |
|   toggle-label    |     position     |               |
|    toggle-span    |     z-index      |               |
|  toggle-wrapper   | background-color |               |
|  toggle-wrapper   |  border-radius   |               |
|  toggle-wrapper   |      height      |               |
