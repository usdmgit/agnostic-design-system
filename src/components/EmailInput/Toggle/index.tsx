import React from 'react';
import Button from '@/components/Button';
import classNames from 'classnames';
import styles from './Toggle.css';

interface Props<T> {
  activeItem: T;
  items: Array<T>;
  variablesClassName?: string;

  getLabel: (item: T) => string;
  getValue: (item: T) => string;
  onChange: (value: T) => void;
}

const Toggle = <T extends {}>(props: Props<T>) => {
  const { items, activeItem, onChange, variablesClassName, getLabel, getValue } = props;
  const buttonWidth = 100 / items.length;
  const calculateTranslation = items.indexOf(activeItem) * 100;

  return (
    <div className={classNames(styles['toggle-wrapper'], variablesClassName)}>
      <div data-testid='toggle-element' className={classNames(styles.toggle, variablesClassName)}>
        <div
          className={classNames(styles['toggle-active-btn'], variablesClassName)}
          style={{
            width: `${buttonWidth}%`,
            transform: `translateX(${calculateTranslation}%)`
          }}
          data-active-item={getValue(activeItem)}
        />
        {items.map(item => {
          return (
            <label
              className={classNames(styles['toggle-label'], variablesClassName)}
              style={{ width: `${buttonWidth}%` }}
              key={getLabel(item)}
              htmlFor={getLabel(item)}
              data-active={activeItem === item}
              data-testid={`toggle-label-${getLabel(item)}`}
            >
              <Button
                size='medium'
                text={getLabel(item)}
                category='neutral'
                onClick={() => onChange(item)}
                data-testid={`toggle-element-${getLabel(item)}`}
                variablesClassName={classNames(styles['toggle-button'], variablesClassName)}
              />
              <input
                readOnly
                id={getLabel(item)}
                type='radio'
                name='toggle'
                checked={activeItem === item}
                value={getValue(item)}
                data-testid={`toggle-input-${getLabel(item)}`}
                className={classNames(styles['toggle-input'])}
              />
            </label>
          );
        })}
      </div>
    </div>
  );
};

export default Toggle;

Toggle.defaultProps = {
  items: [],
  activeItem: null
};
