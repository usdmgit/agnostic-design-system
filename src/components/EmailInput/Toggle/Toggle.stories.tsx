import React, { useState } from 'react';
import Toggle from '@/components/Toggle';
import { Meta } from '@storybook/react';
import mdx from './Toggle.stories.mdx';
import classNames from 'classnames';
import styles from './Toggle.stories.css';

export default {
  title: 'Components/Toggle',
  component: Toggle,
  parameters: {
    docs: {
      page: mdx
    }
  }
} as Meta;

const Template = args => {
  const [active, setActive] = useState(args.activeItem);

  return <Toggle {...args} activeItem={active} onChange={setActive} />;
};

const items = [
  {
    title: 'First'
  },
  {
    title: 'Second'
  }
];

export const Default = Template.bind({});

Default.args = {
  activeItem: items[0],
  getLabel: item => item.title,
  getValue: item => item.title,
  items: items
};

const fourItems = [
  {
    title: 'First'
  },
  {
    title: 'Second'
  },
  {
    title: 'Third'
  },
  {
    title: 'Forth'
  }
];

export const MultipleItems = Template.bind({});

MultipleItems.args = {
  activeItem: fourItems[0],
  getLabel: item => item.title,
  getValue: item => item.title,
  items: fourItems
};

export const customizedToggle = Template.bind({});

customizedToggle.args = {
  activeItem: items[0],
  getLabel: item => item.title,
  getValue: item => item.title,
  items: items,
  variablesClassName: classNames(styles['custom-toggle'])
};
