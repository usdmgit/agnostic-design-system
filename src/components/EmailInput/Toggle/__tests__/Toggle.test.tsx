import React, { useState } from 'react';
import Toggle from '@/components/Toggle';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('Toggle', () => {
  it('renders correctly', () => {
    const wrapper = renderToggle();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Toggle rendered options', () => {
  it('will have the content passed in items', () => {
    renderToggle();
    const toggleRendered = screen.getByTestId('toggle-element');
    expect(toggleRendered).toHaveTextContent(/First/);
    expect(toggleRendered).toHaveTextContent(/Second/);
  });
});

describe('Toggle behaviour', () => {
  it('toggles value when clicked', () => {
    renderToggle();
    const testLabel = screen.getByTestId('toggle-label-First');
    const escalationLabel = screen.getByTestId('toggle-label-Second');

    expect(testLabel.getAttribute('data-active')).toEqual('true');
    expect(escalationLabel.getAttribute('data-active')).toEqual('false');

    fireEvent.click(screen.getByTestId('toggle-element-Second'));

    expect(testLabel.getAttribute('data-active')).toEqual('false');
    expect(escalationLabel.getAttribute('data-active')).toEqual('true');
  });
});

const items = [
  {
    title: 'First'
  },
  {
    title: 'Second'
  }
];

function Wrapper(props) {
  const [active, setActive] = useState(props.active);
  const getLabel = item => {
    return item.title;
  };
  const getValue = item => {
    return item.title;
  };
  return (
    <Toggle
      items={items}
      getLabel={getLabel}
      getValue={getValue}
      activeItem={active}
      onChange={setActive}
    />
  );
}

function renderToggle(activeItem = items[0]) {
  return render(<Wrapper active={activeItem} />);
}
