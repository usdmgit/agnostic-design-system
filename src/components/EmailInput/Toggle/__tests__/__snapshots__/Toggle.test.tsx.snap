// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="toggle-wrapper"
      >
        <div
          class="toggle"
          data-testid="toggle-element"
        >
          <div
            class="toggle-active-btn"
            data-active-item="First"
            style="width: 50%; transform: translateX(0%);"
          />
          <label
            class="toggle-label"
            data-active="true"
            data-testid="toggle-label-First"
            for="First"
            style="width: 50%;"
          >
            <button
              category="neutral"
              class="button button--neutral button button--medium toggle-button"
              data-testid="toggle-element-First"
              type="button"
            >
              <span
                class="button-text"
              >
                First
              </span>
            </button>
            <input
              checked=""
              class="toggle-input"
              data-testid="toggle-input-First"
              id="First"
              name="toggle"
              readonly=""
              type="radio"
              value="First"
            />
          </label>
          <label
            class="toggle-label"
            data-active="false"
            data-testid="toggle-label-Second"
            for="Second"
            style="width: 50%;"
          >
            <button
              category="neutral"
              class="button button--neutral button button--medium toggle-button"
              data-testid="toggle-element-Second"
              type="button"
            >
              <span
                class="button-text"
              >
                Second
              </span>
            </button>
            <input
              class="toggle-input"
              data-testid="toggle-input-Second"
              id="Second"
              name="toggle"
              readonly=""
              type="radio"
              value="Second"
            />
          </label>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="toggle-wrapper"
    >
      <div
        class="toggle"
        data-testid="toggle-element"
      >
        <div
          class="toggle-active-btn"
          data-active-item="First"
          style="width: 50%; transform: translateX(0%);"
        />
        <label
          class="toggle-label"
          data-active="true"
          data-testid="toggle-label-First"
          for="First"
          style="width: 50%;"
        >
          <button
            category="neutral"
            class="button button--neutral button button--medium toggle-button"
            data-testid="toggle-element-First"
            type="button"
          >
            <span
              class="button-text"
            >
              First
            </span>
          </button>
          <input
            checked=""
            class="toggle-input"
            data-testid="toggle-input-First"
            id="First"
            name="toggle"
            readonly=""
            type="radio"
            value="First"
          />
        </label>
        <label
          class="toggle-label"
          data-active="false"
          data-testid="toggle-label-Second"
          for="Second"
          style="width: 50%;"
        >
          <button
            category="neutral"
            class="button button--neutral button button--medium toggle-button"
            data-testid="toggle-element-Second"
            type="button"
          >
            <span
              class="button-text"
            >
              Second
            </span>
          </button>
          <input
            class="toggle-input"
            data-testid="toggle-input-Second"
            id="Second"
            name="toggle"
            readonly=""
            type="radio"
            value="Second"
          />
        </label>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
